### ISSUE TITLE
Question: UNNotificationServiceExtension support on iOS?

### ISSUE THREAD

**Issue by tamimattafi** [mirzemehdi/KMPNotifier#24]
Hello! 
Thank you for this amazing library! 

I'm wondering, is there a possibility we could have support for changing push notification content for iOS? In a native app, it's usually done using `UNNotificationServiceExtension`, is it technically possible we could have access to such functionality directly from the library and on the Kotlin side (iosMain for instance) without having to implement such extension in every project?

**Comment by mirzemehdi** [comment]
How it works in ios? You can change content even app is in background, and after you get push notification you can change content?

**Comment by tamimattafi** [comment]
@mirzemehdi Yes, the extension service is meant to work in the background

**Comment by mirzemehdi** [comment]
However when app is in background, and when receiving notification type of message, notification title and body will be shown as it is to the user as this will be handled by the system, in android it is like that at least.

But you can hack around by sending only data type of message, then based on that you can show to the user whatever you want.

**Comment by tamimattafi** [comment]
@mirzemehdi Is there a possibility of showing a notification from the background, using data type messages? I would be very thankful if you point out to an example.
Unfortunately, I have tried something similar, and it works only if the app is in the foreground

**Comment by mirzemehdi** [comment]
@tamimattafi Looks like using data type messages with ios when it is in background is tricky. For my case, I can get notification type of message both in background and foreground, but I also have a problem receiving data type of message when app is in background.

I also checked this one https://stackoverflow.com/questions/38166203/ios-data-notifications-with-fcm, and even setting `content_available: true` wasn't working for me.

If you can find another solution, you can let us know here as well!

**Comment by tamimattafi** [comment]
@mirzemehdi I implemented `UNNotificationServiceExtension` in my project, it has the following behavior and limitations:
1. It must be in a separate target (as extension), which means it needs its own bundle ID and if you want to access data stored by the main app, you will need to use groups, also the runtime is limited to 24MB. This is very different from what we used to in android
2. When the app is in the background, `UNNotificationServiceExtension` is triggered, it gives you the ability and some time to mutate the original one that came from the server, however, for some cases this might not be enough, for example, without a permission from Apple you can't ignore notifications that you don't want to display, for that, you need to set `mutable-content: true` in your `aps` payload, and you must have an `alert` 
3. When the app is in the foreground, `UNNotificationServiceExtension` is ignored, instead `didReceiveRemoteNotification` is called with `userInfo`. This is a very weird approach for an android developer, we used to have a single entry point.
4. In android, or using `localNotifier`, you can set an `id` to your notification to be able to control it later, but that's not the case for background iOS notifications, the identifier can be set only using `collapse-id`, which must be inside your `aps` payload
5. You might need to set `UNUserNotificationCenter.delegate` in your `didFinishLaunchingWithOptions`, because some push notifications can be ignored and they must be controlled using `func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification)`
6. If your app is very large, you won't be able to use `shared` or `composeApp` in the notification extension (different target) due to memory limitations (24MB), you might need to break down your logic and create a new `shared` module with minimal dependencies, otherwise, the system can kill this process very unexpectedly


My experience with Push Notifications on iOS is kind of interesting but frustrating at the same time, not sure why they chose to have different entry points and different targets. `Service` and `process` systems on Android seem to me more simple and straight forward.

**Comment by mirzemehdi** [comment]
@tamimattafi Thanks for detailed explanation. Look like very complicated thing :D. I need to research that for understanding better

**Comment by Koushiktrimble** [comment]
Hey Guys,
I am from android background. Currently working on a KMP project and am using KMPNotifier library in it. I don't understand your above conversations about iOS. Currently I am trying to send a data type message to my app. But It is not getting received inside onPayloadData method. Is that the known issue? Do we have a workaround for that?

**Comment by koushikvkk** [comment]
Hey Guys,
I am from android background. Currently working on a KMP project and am using KMPNotifier library in it. I don't understand your above conversations about iOS. Currently I am trying to send a data type message to my app. But It is not getting received inside onPayloadData method. Is that the known issue? Do we have any workaround for that?

**Comment by mirzemehdi** [comment]
@koushikvkk yeah this is currently issue, trying to work on that

**Comment by Koushiktrimble** [comment]
Ok. Thanks for the info @mirzemehdi.

**Comment by koushikvkk** [comment]
Ok. Thanks for the info @mirzemehdi.

**Comment by mirzemehdi** [comment]
@koushikvkk for receiving data type message in ios, I wrote here https://github.com/mirzemehdi/KMPNotifier/issues/54#issuecomment-2269773896

Basically main thing that is working for me to get data type only message in ios, is to add `"badge"` key to `aps` when sending notification

**Comment by Koushiktrimble** [comment]
Hey @mirzemehdi. Thanks for the info.
When I tried this, I could see data notification is getting received inside the onPayloadData callback only when the app is in foreground. When the app is in background, the data is not getting received. Do you also face the same issue?

**Comment by mirzemehdi** [comment]
@Koushiktrimble yes, and no. So here is what happens when app is in background. I think this is limitation from ios side, but not sure if someone can find more information let us know please.


1) First case: Sending only data type message (app in background):
In this case when app is in background, no listener is triggered.

2) Second case: Sending both data and notification message (app in background)
In this case, no listener is triggered, but notification is shown to the user. Then when user clicks the notification, data type message is received as well, and listener is also triggered.  

3) Third case: Sending only data type message (app in foreground)
No issue, Listener is triggered, and data type message is received.

4) Fourth case: Sending both data and notification message (app in foreground)
No issue, Listeners are triggered, data type message is received, and notification is shown to the user.

**Comment by tamimattafi** [comment]
@mirzemehdi Hello!
Data notifications are always received even if the app is in the background. The limitations are related to UI, you can't show anything from the background, only fetch data. So, data notifications should be something like silent or background notifications for data fetching or changing something in user cache. Never for UI.
 
If you want to guarantee that the user sees a notification (Display it on the UI), you should always send alerts with badge and sound properties. If you want to customize their text and style on the client side, you should implement Notification Service Extension, and add mutable-content: 1 to your aps payload.

Clicking on a notification should trigger delegate function didFinishLaunchingWithOptions, with a dictionary of data from the push notification you clicked, so you should solely rely on this when you handle clicks and never rely on the data notification.

**Comment by mirzemehdi** [comment]
@tamimattafi thank you for this info. Even printing something to the logs doesn't work. is this expected then?

**Comment by Louisnil-AS** [comment]
@mirzemehdi  @tamimattafi  
May i ask about iOS, is it possible to execute code like simply NSLog a message by this method (UNNotificationServiceExtension). Even the app is fully closed by user (app is killed). So by using UNNotificationServiceExtension I can finally execute code like storing the push message data into my phone even if the app is fully closed? like closed by user?

**Comment by tamimattafi** [comment]
@Louisnil-AS You can write `NSLog` but it won't appear in your logs, because extensions run on a different process, you will need to switch your logs view to listen to this process. 
You can download things like a picture from a URL even if the app is closed, so I assume you can store things to local storage too, but data fetching and storing is recommended using data type notifications and not alerts.

Check these docs: https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension/

**Comment by Louisnil-AS** [comment]
@tamimattafi  
Yes, because normally we can't run code when the app is killed. Like iOS doesn't allow that because of privacy issues... 
Did you mean this Extension can solve the problem?

**Comment by tamimattafi** [comment]
@Louisnil-AS Yes, the Extension runs and allows HTTP requests and local storage operations. But the time should not exceed few seconds, and the total memory consumption of this process should not exceed 24MB. So if you can live with these limitations, it's possible to do what you are looking for.

**Comment by tmdgh1592** [comment]
@mirzemehdi 
hello!
Are you talking about an issue where the onPayloadData() function is not called in the background?
I am wondering if it is possible to have the onPayloadData() function be called when fcm is received in the background.

**Comment by mirzemehdi** [comment]
@tmdgh1592 onPayloadData functions is called in background if it contains both notification and data type message. But if it contains data type message (silent message), then it is not called. But check chis comment too: https://github.com/mirzemehdi/KMPNotifier/issues/24#issuecomment-2272496556

**Comment by tmdgh1592** [comment]
@mirzemehdi Thank you
I tried data + notification as you said.
If you include notification, it is true that push notifications are sent in the background, but the onPayloadData() function is still not called.

I tried it in postman
```
// url : https://fcm.googleapis.com/v1/projects/{projectId}/messages:send

{
  "message": {
    "token": "eCGO2Cbmf...",
    "data": {
      "customKey1":"...",
      "customKey2":"..."
    },
    "notification": {
        "title":"ddd",
        "body":"asdasa"
    }
  }
}
```

I checked the comments and tried including apns and badge instead of notification type,
In the same way, only push notifications are sent, and onPayloadData() is not called.
Is there perhaps a problem with the way I tried?

**Comment by mirzemehdi** [comment]
Correction to above request @tmdgh1592  would be adding apns -> .... badge also as below

```

        "message": {
            "token": device_id,
            "notification": {
                "title": "FCM Notification",
                 "body": "Notification from FCM"
            },
            "data": data,
            "apns": {
                "payload": {
                    "aps": {
                        "badge": 0,
                    },
                }
            }
        }
    

```

**Comment by tmdgh1592** [comment]
@mirzemehdi 
Does device_id mean fcm token??

I tried this format and the result is still the same
I tried logging, but I dont see anything 🥲

**Comment by mirzemehdi** [comment]
@tmdgh1592 yes device_id is fcm token. in logs you will not see payload data when app is in background. But when you click notification, then after app comes foreground onPayloadData will be triggered and you will get data there

**Comment by tmdgh1592** [comment]
@mirzemehdi 
I thought this was the case after reading the comments you left, but onPayloadData() is not called even when I click on the notification.
There is no problem at all when in the foreground, but this problem only occurs in the background.

**Comment by mirzemehdi** [comment]
@tmdgh1592  then make sure in ios
you call `NotifierManager.onApplicationDidReceiveRemoteNotification(userInfo: userInfo)` on application's `didReceiveRemoteNotification` method. 

https://github.com/mirzemehdi/KMPNotifier?tab=readme-ov-file#ios

If you did that step too and all other steps in ReadMe, not sure what can be missing

**Comment by tmdgh1592** [comment]
@mirzemehdi
yes, I already called `onApplicationDidReceiveRemoteNotification()` and did steps in ReadMe

```swift
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any]) async -> UIBackgroundFetchResult {
        NotifierManager.shared.onApplicationDidReceiveRemoteNotification(userInfo: userInfo)
        return UIBackgroundFetchResult.newData
    }
```

```swift
import SwiftUI
import ComposeApp
import FirebaseCore
import FirebaseMessaging

class AppDelegate: NSObject, UIApplicationDelegate {

    func application(_ application: UIApplication,
                   didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool {
        FirebaseApp.configure()
        NotifierInitializer.shared.onApplicationStart()
      
        return true
    }

    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
        Messaging.messaging().apnsToken = deviceToken
    }
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any]) async -> UIBackgroundFetchResult {
        NotifierManager.shared.onApplicationDidReceiveRemoteNotification(userInfo: userInfo)
        return UIBackgroundFetchResult.newData
    }
    
}

@main
struct iOSApp: App {
    
    @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate
    
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}
```

As of now, I don't really know...
thank you for your help

**Comment by tamimattafi** [comment]
@tmdgh1592 Have you tried adding this flag? `'content-available': 1`
Check this thread https://stackoverflow.com/questions/27777632/the-purpose-of-content-available-in-push-notification-json

**Comment by tmdgh1592** [comment]
@tamimattafi 

I have implemented a function that shows push notifications in onPayloadData().
If you set content_available, fcm is received in the background, but the logic of onPayloadData() is not called because no push notifications are visible.

I tried like below

```
{
  "message": {
    "token": "eSMzRHZ7A0XSlR8r-...",
    "data": data,
    "apns": {
      "payload": {
        "aps": {
            "content-available": 1
        }
      }
    }
  }
}
```

**Comment by tamimattafi** [comment]
@tmdgh1592 I don't use `onPayloadData` to show notifications, I use [UNNotificationServiceExtension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension/) with `'mutable-content': 1`, so my payload is irrelevant for your experiment :(

**Comment by tmdgh1592** [comment]
@tamimattafi Still, thank you for your helpful comment, bro.

**Comment by MahdiAsoudeh** [comment]
Hello Mr. @mirzemehdi 
thanks a lot for this library, winderful
but I have an error in iOS yet, please help me, god save you

![2](https://github.com/user-attachments/assets/c36a879a-30d2-4b36-b649-45f909d356ea)

![1](https://github.com/user-attachments/assets/b1a76ca4-70e5-430a-b206-05cc4f8959ed)

**Comment by LucasPrioste92** [comment]
@tmdgh1592 I'm facing the same issue — I'm not receiving the payload when the app is in the background, and the user taps the notification. Did you manage to find a solution?
